FW/Presentations/Lab_Guides/_L_C_04_Writing_UEFI_App_Simics_Linux_Lab.md

```diff
@@ -0,0 +1,2024 @@
+3. Next Update the SampleApp.inf to include: SimicsOpenBoardPkg/OpenBoard.dec, the new PcD, and the PcdLib (hint: see how HelloWorld.inf used the PCD enable)
+4. SampleApp.c can now use the newly created PCD 
+
+---
+## Slide 60 Lab 5: Solution: Dec & Inf
+
+**MyPkg/MyPkg.dec**
+
+```
+[PcdsFeatureFlag]
+ ## Indicates if SampleApp Application will enable the Typewriter Feature.
+  #  This PCD is a sample to explain FeatureFlag PCD usage.<BR><BR>
+  #   TRUE  - SampleApp Application will enable the Typewriter Feature.<BR>
+  #   FALSE - SampleApp Application will not enable the Typewriter Feature.<BR>
+  # @Prompt Enable SampleApp enable the Typewriter Feature.
+ gEfiMyPkgTokenSpaceGuid.PcdTypeWriterFeatureEnable|TRUE|BOOLEAN|0x0001200a
+```
+
+**SampleApp.inf**
+
+```
+. . .
+[Packages]
+  MdePkg/MdePkg.dec
+  MyPkg/MyPkg.dec
+  
+[LibraryClasses]
+   UefiBootServicesTableLib
+   UefiApplicationEntryPoint
+   UefiLib
+   DebugLib
+   PcdLib
+
+[FeaturePcd]
+   gEfiMyPkgTokenSpaceGuid.PcdTypeWriterFeatureEnable  ## CONSUMES
+```
+
+---
+## Slide 61 Lab 5: Solution
+
+(hint: Lesson B.5 has the solution)
+
+**SampleApp.c**
+
+```c
+#include <Uefi.h>
+#include <Library/UefiApplicationEntryPoint.h>
+#include <Library/UefiLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#define  CHAR_DOT  0x002E    // '.' in Unicode
+
+EFI_STATUS
+EFIAPI
+UefiMain (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  UINTN          EventIndex;
+  BOOLEAN        ExitLoop;
+  EFI_INPUT_KEY  Key;
+  
+// Lab 3
+ Print(L"System Table: 0xp/n",SystemTable); 
+
+//Lab 4
+ Print( L"/nPress any Key to continue : /n");
+ gBS->WaitForEvent (1, &gST->ConIn->WaitForKey,EventIndex);
+
+// Lab 5
+ gST->ConIn->ReadKeyStroke (gST->ConIn, &Key);
+ if (FeaturePcdGet(PcdTypeWriterFeatureEnable)) {
+   Print(L"Enter text. Include a dot ('.') in a /
+     sentence then <Enter> to exit:/n/n");
+   ZeroMem (&Key, sizeof (EFI_INPUT_KEY));
+   ExitLoop = FALSE;
+   do {
+   gBS->WaitForEvent (1, &gST->ConIn->WaitForKey,       &EventIndex);
+   gST->ConIn->ReadKeyStroke (gST->ConIn, &Key);
+   Print(L"%c", Key.UnicodeChar);
+   if (Key.UnicodeChar ==  CHAR_DOT){
+    ExitLoop = TRUE;
+       }
+     } while (!(Key.UnicodeChar == CHAR_CARRIAGE_RETURN) || 
+              !(ExitLoop) );
+ }
+ Print(L"/n");
+ return EFI_SUCCESS;
+}
+```
+
+
+Copy and Paste from LabGuide (This document)
+
+---
+## Slide 62 Lab 5: Build and Test SampleApp
+
+1. At the Terminal Command Prompt, Re-Build BoardX58Ich10
+
+```bash
+$> cd ~/fw/edk2-ws/edk2-platforms/Platform/Intel/
+$> python build_bios.py -p BoardX58Ich10 -t GCC5
+```
+
+2. Copy SampleApp.efi from the build directory to the VHD Disk Copy `../Build/SimicsOpenBoardPkg/BoardX58Ich10/DEBUG_GCC5/X64/SampleApp.efi UefiAppLab Disk`
```

Comment:
2. Copy SampleApp.efi from the build directory to the VHD Disk Copy 
`$ ../Build/SimicsOpenBoardPkg/BoardX58Ich10/DEBUG_GCC5/X64/SampleApp.efi UefiAppLab ~/VHD`